docs/content/contributing/monorepo.md

```diff
@@ -0,0 +1,103 @@
+---
+
+# Monorepo Structure
+
+The [kcp-dev/kcp](https://github.com/kcp-dev/kcp) repository is a monorepo that contains the source code for:
+
+- the kcp core
+- libraries that are used in the kcp core or are close to the kcp core
+
+At the moment, the following libraries are part of the monorepo:
+
+- [kcp-dev/apimachinery](https://github.com/kcp-dev/apimachinery)
+- [kcp-dev/client-go](https://github.com/kcp-dev/client-go)
+- [kcp-dev/code-generator](https://github.com/kcp-dev/code-generator)
+
+## Terminology
+
+- The monorepo: the [kcp-dev/kcp](https://github.com/kcp-dev/kcp) repository
+- Staging repository: a repository that is part of the monorepo
+  (located within `kcp-dev/kcp/staging/src/github.com/kcp-dev`)
+- Published repository: a standalone repository that's a mirror of the staging repository
+
+Note: the term 'staging repository' does **not** imply a Git repository (with `.git` directory), but an
+authoritative place that stores code. Usually, each staging repository is a standalone Go module.
+
+## Similarities with Kubernetes
+
+This pattern is the same as the Kubernetes staging repositories pattern, with an exception of how tags are handled
+within published repositories (this is explained later in this document).  
+
+## High Level Overview
+
+The source code for staging repositories is located on the special path in the monorepo which is
+[`./staging/src/github.com/kcp-dev`](https://github.com/kcp-dev/kcp/tree/main/staging/src/github.com/kcp-dev).
+Every subdirectory on that path corresponds to a single staging repository.
+The code in the staging/ directory is authoritative, i.e. the only copy of the code. You can directly modify such code.
+
+In order to preserve standalone Go modules for each staging repository, the source code from staging repositories
+is mirrored to standalone (aka published) repositories. The mirroring is an automatic process happening every
+several hours.
+
+Aside from the source code, the mirroring process includes branches and tags. This means that branches from
+the monorepo for which mirroring is enabled will be created in each published repository. Tags on those branches
+will be automatically mirrored to the published repositories. At the moment, mirroring is enabled for
+the `main` branch and release branches starting from `release-0.29`.
+
+## Handling Tags
+
+Tags published to the published repositories inherit the major version of their original Go module. That means if
+the Go module major version for that library is v2, the tag will have v2 as the major version.
+For example, if we push v0.29.0 in the kcp repository, the tag in the published repository that has Go module major
+version v2 will be v2.29.0.
+
+Concretely:
+
+- kcp-dev/client-go has the Go module major version at v0, so tags will be v0.x.y
+- kcp-dev/apimachinery has the Go module major version at v2, so tags will be v2.x.y
+- kcp-dev/code-generator has the Go module major version at v3, so tags will be v3.x.y
+
```

Comment:
It's worth noting that these tags are unrelated to actual versions of kcp; they are V2, V3, to maintain backwards compatibility with the pre-monorepo tagging strategy.